source/extensions/filters/network/thrift_proxy/router/router_impl.cc

```diff
@@ -522,9 +522,22 @@ void Router::UpstreamRequest::onRequestStart(bool continue_decoding) {
   }
 }
 
-void Router::UpstreamRequest::onRequestComplete() { request_complete_ = true; }
+void Router::UpstreamRequest::onRequestComplete() {
+  Event::Dispatcher& dispatcher = parent_.callbacks_->dispatcher();
+  downstream_request_complete_time_ = dispatcher.timeSource().monotonicTime();
+  request_complete_ = true;
+}
 
 void Router::UpstreamRequest::onResponseComplete() {
+  if (!charged_response_timing_) {
+    Event::Dispatcher& dispatcher = parent_.callbacks_->dispatcher();
+    const std::chrono::milliseconds response_time =
+        std::chrono::duration_cast<std::chrono::milliseconds>(
+            dispatcher.timeSource().monotonicTime() - downstream_request_complete_time_);
+    parent_.chargeResponseTiming(response_time);
+    charged_response_timing_ = true;
+  }
```

Comment:
nit: move this block -- with the if guard included -- to `chargeResponseTiming()` and call it here and below.